EpsilonAuxBMS/Tests/tests/UpdateChargeAllowanceTaskTest/UpdateChargeAllowanceTaskTest.c

```diff
@@ -0,0 +1,178 @@
+    RUN_TEST(test_UpdateChargeAllowanceTask_TooHighMaxCellVoltageShouldTurnOffChargeAndUpdateStatuses);
+    RUN_TEST(test_UpdateChargeAllowanceTask_TooLowMinCellVoltageShouldTurnOffDischargeAndUpdateStatuses);
+}
+
+void setUpdateChargeAllowanceOsExpects(uint32_t* prevWakeTime)
+{
+    osDelayUntil_ExpectAndReturn(prevWakeTime, 100, osOK);
+    osMutexWait_ExpectAndReturn(auxStatus.auxStatusMutex, 0, osOK);
+    osMutexRelease_ExpectAndReturn(auxStatus.auxStatusMutex, osOK);
+    osMutexWait_ExpectAndReturn(orionStatus.orionStatusMutex, 0, osOK);
+    osMutexRelease_ExpectAndReturn(orionStatus.orionStatusMutex, osOK);
+}
+
+void setOrionEnableSenseExpects(GPIO_PinState expectedOrionDischargeSense, GPIO_PinState expectedOrionChargeSense)
+{
+    auxStatus.startUpSequenceDone = 1;
+    HAL_GPIO_ReadPin_ExpectAndReturn(ORION_DISCHARGE_ENABLE_SENSE_GPIO_Port, ORION_DISCHARGE_ENABLE_SENSE_Pin, expectedOrionDischargeSense);
+    HAL_GPIO_ReadPin_ExpectAndReturn(ORION_CHARGE_ENABLE_SENSE_GPIO_Port, ORION_CHARGE_ENABLE_SENSE_Pin, expectedOrionChargeSense);
+}
+
+void test_UpdateChargeAllowanceTask_StandardOperationShouldUpdateAuxAndOrionStatusCorrectly()
+{
+    uint32_t prevWakeTime = 0;
+    setUpdateChargeAllowanceOsExpects(&prevWakeTime);
+    GPIO_PinState expectedOrionDischargeSense = GPIO_PIN_SET;
+    GPIO_PinState expectedOrionChargeSense = GPIO_PIN_SET;
+
+    setOrionEnableSenseExpects(expectedOrionDischargeSense, expectedOrionChargeSense);
+
+    carShouldTrip_ExpectAndReturn(0);
+
+    orionStatus.canMsgReceived = 1;
+    orionStatus.maxCellVoltage = 30000;
+    orionStatus.minCellVoltage = 30000;
+    updateChargeAllowance(&prevWakeTime);
+
+    TEST_ASSERT_EQUAL(1, orionStatus.allowDischarge);
+    TEST_ASSERT_EQUAL(1, orionStatus.allowCharge);
+    TEST_ASSERT_EQUAL(1, orionStatus.batteryVoltagesInRange);
+    TEST_ASSERT_EQUAL(0, orionStatus.shutOff);
+    TEST_ASSERT_EQUAL(0, orionStatus.chargeContactorOverriden);
+    TEST_ASSERT_EQUAL(0, orionStatus.dischargeContactorOverriden);
+
+    TEST_ASSERT_EQUAL(0, auxStatus.strobeBmsLight);
+    TEST_ASSERT_EQUAL(1, auxStatus.allowCharge);
+}
+
+void test_UpdateChargeAllowanceTask_NoDischargeSenseShouldTurnOffPinAndUpdateOrionStatus()
+{
+    uint32_t prevWakeTime = 0;
+    setUpdateChargeAllowanceOsExpects(&prevWakeTime);
+
+    GPIO_PinState expectedOrionDischargeSense = GPIO_PIN_RESET;
+    GPIO_PinState expectedOrionChargeSense = GPIO_PIN_SET;
+    setOrionEnableSenseExpects(expectedOrionDischargeSense, expectedOrionChargeSense);
+
+    carShouldTrip_ExpectAndReturn(0);
+
+    // Test if PIN discharge enable and HV was written to
+    HAL_GPIO_WritePin_Expect(HV_ENABLE_GPIO_Port, HV_ENABLE_Pin, GPIO_PIN_RESET);
+    HAL_GPIO_WritePin_Expect(DISCHARGE_ENABLE_GPIO_Port, DISCHARGE_ENABLE_Pin, GPIO_PIN_RESET);
+    updateChargeAllowance(&prevWakeTime);
+
+    TEST_ASSERT_EQUAL(0, orionStatus.allowDischarge);
+    TEST_ASSERT_EQUAL(1, orionStatus.dischargeContactorOverriden);
```

Comment:
Should probably also check that the charge related members are what is expected as well

Reply:
My thinking was that `test_UpdateChargeAllowanceTask_NoDischargeSenseShouldTurnOffPinAndUpdateOrionStatus` only cares about the discharge contactor, making sure that no matter what occurs with the charge sense, as long as the orionStatus updated the discharge stuff it was ok. I guess the downside is that I'm not checking if something unintentionally mutates the other orionStatus fields too.

Gonna try making it expect that the orionStatus doesn't change, except for certain fields.

Reply:
Yeah that was my thinking. It's probably safer to check that nothing else changes unexpectedly.